crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1673,6 +1673,128 @@ impl<'db> Type<'db> {
+                // ≈ A.exclude_always_truthy() & ~B
+                //
+                // Case 2: Positives are empty
+                // ~B & Falsy
+                // = ~(B | AlwaysTruthy)
+                // ≈ ~(B.include_always_truthy())
+                //
+                // Note: Calculating `(~B & Falsy)` = `~(B | AlwaysTruthy)` is extremely difficult.
+                // This set should include all elements where `bool(element)` is always True,
+                // regardless of the type of B. Examples include:
+                // { *(elements of B), True, 1, 2, ..., -1, -2, ..., *(instances of object), ... }.
+                //
+                // Attempting to represent this set via Union or Intersection
+                // could break type inference.
+                let mut builder = IntersectionBuilder::new(db);
+
+                // Theoretically, applying `intersection` to a single positive or to all positives
+                // should yield the same result. However, due to the limitations of the current type system,
+                // there is a possibility of losing information when applying it to only one positive.
+                // As a result, `exclude_always_truthy` is applied to all positives.
+                for element in intersection.positive(db) {
+                    builder = builder.add_positive(element.exclude_always_truthy(db));
+                }
+
+                for element in intersection.negative(db) {
+                    builder = builder.add_negative(*element);
+                }
+
+                debug_assert!(
+                    !intersection.positive(db).is_empty(),
+                    "negative-only-intersections can not use `exclude_always_truthy`"
+                );
+
+                builder.build()
+            }
+            Type::Union(union) => {
+                let mut builder = UnionBuilder::new(db);
+
+                for element in union.elements(db) {
+                    builder = builder.add(element.exclude_always_truthy(db));
+                }
+
+                builder.build()
+            }
+            // Since `bool` is a @final class, there is no need to worry about subclasses.
+            Type::Instance(instance) if instance.class.is_known(db, KnownClass::Bool) => {
+                Type::BooleanLiteral(false)
+            }
+            _ => {
+                if let Truthiness::AlwaysTrue = self.bool(db) {
+                    return Type::Never;
+                }
+
+                self
+            }
+        }
+    }
+
+    pub fn falsy_literals(db: &'db dyn Db) -> Type<'db> {
+        UnionBuilder::new(db)
+            .add(Type::none(db))
+            .add(Type::BooleanLiteral(false))
+            .add(Type::IntLiteral(0))
+            .add(Type::string_literal(db, ""))
+            .add(Type::bytes_literal(db, b""))
+            // TODO: tuple literal should be included
+            // .add(Type::tuple(db, &[]))
+            .build()
+    }
```

Comment:
In the vast majority of cases, you want to use `UnionType::from_elements()` or `UnionBuilder` when you're constructing a union, like you do here. The reason why is that it's a smart constructor that takes care to deduplicate elements and remove overlapping subtypes before creating the `UnionType` object. However, in this specific case, we _know_ the types are not overlapping, so we can avoid some indirection here, and use `UnionType::new()` directly:

```suggestion
    pub fn falsy_literals(db: &'db dyn Db) -> Type<'db> {
        let elements = Box::from([
            Type::none(db),
            Type::BooleanLiteral(false),
            Type::IntLiteral(0),
            Type::string_literal(db, ""),
            Type::bytes_literal(db, b""),
            // TODO: tuple literal should be included
            // .add(Type::tuple(db, &[]))
        ]);
        Type::Union(UnionType::new(db, elements))
    }
```